src/messaging/ExchangeContext.cpp

```diff
@@ -0,0 +1,390 @@
+ *
+ */
+void ExchangeContext::AddRef()
+{
+    mRefCount++;
+#if defined(CHIP_EXCHANGE_CONTEXT_DETAIL_LOGGING)
+    ChipLogProgress(ExchangeManager, "ec id: %d [%04" PRIX16 "], refCount++: %d", (this - mExchangeMgr->ContextPool + 1),
+                    mExchangeId, mRefCount);
+#endif
+}
+
+void ExchangeContext::DoClose(bool clearRetransTable)
+{
+    // Clear protocol callbacks
+    mDelegate = nullptr;
+
+    // Cancel the response timer.
+    CancelResponseTimer();
+}
+
+/**
+ *  Gracefully close an exchange context. This call decrements the
+ *  reference count and releases the exchange when the reference
+ *  count goes to zero.
+ *
+ */
+void ExchangeContext::Close()
+{
+    VerifyOrDie(mExchangeMgr != nullptr && mRefCount != 0);
+
+#if defined(CHIP_EXCHANGE_CONTEXT_DETAIL_LOGGING)
+    ChipLogProgress(ExchangeManager, "ec id: %d [%04" PRIX16 "], %s", (this - mExchangeMgr->ContextPool + 1), mExchangeId,
+                    __func__);
+#endif
+
+    DoClose(false);
+    Release();
+}
+/**
+ *  Abort the Exchange context immediately and release all
+ *  references to it.
+ *
+ */
+void ExchangeContext::Abort()
+{
+    VerifyOrDie(mExchangeMgr != nullptr && mRefCount != 0);
+
+#if defined(CHIP_EXCHANGE_CONTEXT_DETAIL_LOGGING)
+    ChipLogProgress(ExchangeManager, "ec id: %d [%04" PRIX16 "], %s", (this - mExchangeMgr->ContextPool + 1), mExchangeId,
+                    __func__);
+#endif
+
+    DoClose(true);
+    Release();
+}
+
+/**
+ *  Release reference to this exchange context. If count is down
+ *  to one then close the context, reset all protocol callbacks,
+ *  and stop all timers.
+ *
+ */
+void ExchangeContext::Release()
+{
+    VerifyOrDie(mExchangeMgr != nullptr && mRefCount != 0);
+
+    if (mRefCount == 1)
+    {
+        // Ideally, in this scenario, the retransmit table should
+        // be clear of any outstanding messages for this context and
+        // the boolean parameter passed to DoClose() should not matter.
+        ExchangeManager * em = mExchangeMgr;
+#if defined(CHIP_EXCHANGE_CONTEXT_DETAIL_LOGGING)
+        uint16_t tmpid = mExchangeId;
+#endif
+
+        DoClose(false);
+        mRefCount    = 0;
+        mExchangeMgr = nullptr;
+
+        em->DecrementContextsInUse();
+
+#if defined(CHIP_EXCHANGE_CONTEXT_DETAIL_LOGGING)
+        ChipLogProgress(ExchangeManager, "ec-- id: %d [%04" PRIX16 "], inUse: %d, addr: 0x%x", (this - em->ContextPool + 1), tmpid,
+                        em->GetContextsInUse(), this);
+#endif
+        SYSTEM_STATS_DECREMENT(chip::System::Stats::kExchangeMgr_NumContexts);
+    }
+    else
+    {
+        mRefCount--;
+#if defined(CHIP_EXCHANGE_CONTEXT_DETAIL_LOGGING)
+        ChipLogProgress(ExchangeManager, "ec id: %d [%04" PRIX16 "], refCount--: %d", (this - mExchangeMgr->ContextPool + 1),
+                        mExchangeId, mRefCount);
+#endif
+    }
+}
+
+CHIP_ERROR ExchangeContext::ResendMessage()
+{
+    PayloadHeader payloadHeader;
```

Comment:
When is this function used? Is this resending a previously sent message? If so, the `mMsg` would have already been encrypted, since we encrypt in place. The resend, will try to encrypt it again, which will not work...

Reply:
You are right, this will not work.  Considering CRMP will be enabled by default in CHIP, we don't need re-transmission from protocol level. Removed.